src/Agent/NewRelic/Profiler/Profiler/FunctionResolver.h

```diff
@@ -0,0 +1,200 @@
+/*
+* Copyright 2020 New Relic Corporation. All rights reserved.
+* SPDX-License-Identifier: Apache-2.0
+*/
+#pragma once
+
+#include <map>
+#include "../Logging/Logger.h"
+#include "Function.h"
+#include <cor.h>
+#include <corprof.h>
+
+namespace NewRelic { namespace Profiler
+{
+
+    struct ModuleAndMethodID {
+        ModuleID moduleID;
+        mdMethodDef methodID;
+
+        ModuleAndMethodID(ModuleID moduleID, mdMethodDef methodID)
+        {
+            this->moduleID = moduleID;
+            this->methodID = methodID;
+        }
+
+        bool operator==(const ModuleAndMethodID& other) const
+        {
+            return moduleID == other.moduleID && methodID == other.methodID;
+        }
+
+        bool operator<(const ModuleAndMethodID& other) const
+        {
+            return moduleID < other.moduleID || methodID < other.methodID;
+        }
+    };
+
+    const FunctionID INVALID_FUNCTION_ID = 0;
+
+    // Most of our JIT logic works off of FunctionIDs, but we see method reJITs through the GetReJITParameters
+    // callback which does not give us the id.  Instead it gives us the module and method ids which can map to one or
+    // more FunctionIDs in the case of generic functions.  This resolver class helps us track the mapping between
+    // FunctionIDs and module/method ids for those generic functions.
+    //
+    // If instrumentation is added to a generic method after the agent starts, we find the module id and method id of 
+    // the method and ask for a reJIT.  In this scenario the GetReJITParameters method is called but this class does not
+    // yet have the function id.  GetReJITParameters will finish without adding instrumentation, then ReJITCompilationStarted
+    // will be called with the function id.  That method notifies this class by calling RequestGenericMethodReJIT.
+    // We look up the module/method id and check if it's been queued to be rejitted.  If it is, we add the function id 
+    // info to our map and request another reJIT.  When that triggers, GetReJITParameters will be called once again 
+    // and we will have the function id.
+    // 
+    // Note that the bytecode for a given generic class / method *should* be the same regardless of the specific generic
+    // type as noted by Broman (https://blogs.msdn.microsoft.com/davbr/2011/10/12/rejit-a-how-to-guide/).  For this reason
+    // we only map a single function id to a module/method.  This is okay because when we look up the class and method details
+    // for a function id we use the apis that don't return the type specifics of the generic method.  That means that while 
+    // multiple function ids may map to a single generic method, the AgentShim will see invocations for all of those function ids 
+    // use a single function id.  For our purposes, at least for the current agent functionality we support, this is fine.
+    class FunctionResolver
```

Comment:
We should probably go through this class and determine if any of the method parameters should be made into `const &` params.

Reply:
@alanwest and I talked about this and the parameters are basically ints so we're not concerned with any extra allocations caused by the parameters not being references.